docs/adr/adr-007-monitoring-and-alerting.md

```diff
@@ -0,0 +1,234 @@
+was broken or not. We felt there was room for improvement.
+
+[monitor-time-frame]: https://github.com/alphagov/email-alert-api/blob/c70fcd7e5d64393074a3eebd83b3467fd8cc03b3/app/workers/digest_run_worker.rb#L6-L9
+[monitor-workload]: https://github.com/alphagov/govuk-puppet/blob/fac4b5163f10736bb19aa51a8eb7bc276dd2cb40/modules/govuk/manifests/apps/email_alert_api/sidekiq_queue_check.pp#L33-L44
+[email-alert-monitoring]: https://github.com/alphagov/email-alert-monitoring
+[Travel Advice Publisher]: https://github.com/alphagov/travel-advice-publisher
+
+## Decision
+
+Our decision is that the system should raise a critical alert
+[if and only if](https://en.wikipedia.org/wiki/If_and_only_if) the email alert
+system is broken. The majority of these alerts should contact engineers
+out-of-hours.
+
+Unpacking this, we consider the email alert system to be comprised of 3
+applications: Email Alert API,
+[Email Alert Frontend](https://github.com/alphagov/email-alert-frontend) and
+[Email Alert Service](https://github.com/alphagov/email-alert-service); this
+excludes the monitoring of whether GOV.UK publishing is working and whether
+Notify is working as they are independently monitored. The use of
+"if and only if" reflects a condition we're forming that the system will not
+raise critical alerts unless we're highly confident it is broken, whereas
+warnings are appropriate for other situations. We decided that for the email
+alert system to be "broken" it represents a failing that will lead to a user
+noticeable issue unless an engineer intervenes. Since these alerts should
+represent a broken system requiring intervention, it should be suitable to
+contact engineers out-of-hours for them as otherwise users will be
+experiencing issues.
+
+We decided that we will re-approach our alerting with this mindset and look to
+replace alerts that are vulnerable to reflecting a busy status.
+
+## Status
+
+Pending
+
+## Consequences
+
+We will be changing the approaching to monitoring and alerting across the email
+alert system of applications in order to meet this new alerting criteria.
+
+## Monitoring Email Alert API
+
+### Sidekiq workers
+
+We will configure critical alerts, that contact engineers out-of-hours, if
+individual Sidekiq jobs fail multiple times for the following workers:
+
+- [DailyDigestInitiatorWorker][]
+- [DigestEmailGenerationWorker][]
+- [ProcessContentChangeWorker][]
+- [ProcessMessageWorker][]
+- [SendEmailWorker][]
+- [WeeklyDigestInitiatorWorker][]
+
+These workers will use the database to store the amount of times we have
+attempted to perform an action. An alert will be raised if the quantity of
+failed attempts reaches a threshold (the suggestion is 5).
+
+We will contact engineers out-of-hours if either [SendEmailWorker][] or
+[DigestEmailGenerationWorker][] have not completed at least one successful run
+within a time period (suggestion is 30 minutes) despite work existing. This
```

Comment:
Sure, I can explain. Currently our pattern is to only recover lost work [after an hour](https://github.com/alphagov/email-alert-api/blob/469efe788a290d1a2205d8f870c6aafb4c8c9b00/app/workers/recover_lost_jobs_worker/unprocessed_check.rb#L13). Even if we reduced the that to 30 mins, the worker only runs every half hour, so it could be up to an hour before we recover a lost email. If the total number of emails being sent is small in total (let's say we're processing a "tiny" content change at 11PM), then in the worst case it would be (say) 59 mins between the emails being created and a lost one being recovered and sent.

There's always the possibility of being unlucky, of course. Since we know this happens every so often, I was wondering if we should mention it here as a general caveat: the thresholds may need to be longer (or the recovery more frequent) to account for one or two losses in work.

Does that help?

Reply:
Thanks, I understand now and is a valid point. 

The time period in RecoverLostJobsWorker should ideally be less than the period before we alert and really shouldn't be greater than it. I've modified the ADR to include a reference to this but haven't gone the whole way in defining the solution now - hopefully we can that out when working out how to retry SendEmailWorker (and apply the same to the DigestEmailGenerationWorker).

Reply:
Yes, I agree we can sort this out as an implementation detail.